// @testla/screenplay-playwright/web/actions/Remove.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Remove = void 0;
const screenplay_1 = require("@testla/screenplay");
const BrowseTheWeb_1 = require("../abilities/BrowseTheWeb");
/**
 * Action Class. Remove either Session Storage Items or Local Storage Items from the Browser.
 */
class Remove extends screenplay_1.Action {
    constructor(mode, payload) {
        super();
        this.mode = mode;
        this.payload = payload;
    }
    /**
     * wait for either a specified loading state or for a selector to become visible/active.
     *
     * @param {Actor} actor Actor performing this action
     * @return {any} Returns the value of the `pageFunction` invocation.
     */
    performAs(actor) {
        const { abilityAlias, mode, payload } = this;
        const page = BrowseTheWeb_1.BrowseTheWeb.as(actor, abilityAlias).getPage();
        return page.evaluate(({ k, m }) => {
            const storage = m === 'sessionStorage' ? sessionStorage : localStorage;
            storage.removeItem(k);
            return Promise.resolve();
        }, { k: payload, m: mode });
    }
    /**
     * Remove a session storage item, if a key/value pair with the given key exists.
     *
     * @param {string} key the key that specifies the item.
     * @return {Remove} new Remove instance for session storage
     */
    static sessionStorageItem(key) {
        const instance = new Remove('sessionStorage', key);
        instance.setCallStackInitializeCalledWith({ key });
        return instance;
    }
    /**
     * Remove a local storage item, if a key/value pair with the given key exists.
     *
     * @param {string} key the key that specifies the item.
     * @return {Remove} new Remove instance for local storage
     */
    static localStorageItem(key) {
        const instance = new Remove('localStorage', key);
        instance.setCallStackInitializeCalledWith({ key });
        return instance;
    }
}
exports.Remove = Remove;
