// @testla/screenplay-playwright/api/actions/Head.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Head = void 0;
const UseAPI_1 = require("../abilities/UseAPI");
const constants_1 = require("../constants");
const ARequest_1 = require("./ARequest");
/**
 * Action Class. Send a HTTP HEAD Request.
 */
class Head extends ARequest_1.ARequest {
    constructor(url) {
        super();
        this.url = url;
    }
    /**
     * Send a HTTP HEAD request to the specified url.
     *
     * @param {Actor} actor the actor
     * @return {Response} the response
     */
    performAs(actor) {
        return __awaiter(this, void 0, void 0, function* () {
            return UseAPI_1.UseAPI.as(actor, this.abilityAlias).sendRequest(constants_1.RequestMethod.GET, this.url, this.headers, 'none');
        });
    }
    /**
     * Send a HTTP head request to the specified url.
     *
     * @param {string} url the URL of the target.
     * @return {Head} new instance
     */
    static from(url) {
        const instance = new Head(url);
        instance.setCallStackInitializeCalledWith({ url });
        return instance;
    }
    /**
     * Add headers to the HTTP HEAD request to send.
     *
     * @param {Headers} headers the headers.
     * @return {Head} this instance
     */
    withHeaders(headers) {
        this.headers = headers;
        this.addToCallStack({ caller: 'withHeaders', calledWith: { headers } });
        return this;
    }
}
exports.Head = Head;
