// @testla/screenplay/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ACTIVITY_TYPE = exports.EXEC_STATUS = exports.LOGGING_BLANKS_PER_INDENTATION_LEVEL = exports.LOGGING_BASE_INDENTATION = exports.BASH_COLOR = exports.STRUCTURED_LOGS_ENVVAR_NAME = exports.LOGGING_IDENTIFIER = void 0;
exports.LOGGING_IDENTIFIER = 'testla:sp';
exports.STRUCTURED_LOGS_ENVVAR_NAME = 'TEASLA_SCREENPLAY_STRUCTURED_LOGS';
exports.BASH_COLOR = {
    GRAY: '\x1B[90m',
    BLUE: '\x1B[94m',
    RED: '\x1B[31m',
    RED_BG: '\x1B[41m',
    RESET: '\x1B[0m',
};
// base indentation to match the pw:api indentation
exports.LOGGING_BASE_INDENTATION = '  ';
exports.LOGGING_BLANKS_PER_INDENTATION_LEVEL = 4;
var EXEC_STATUS;
(function (EXEC_STATUS) {
    EXEC_STATUS["START"] = "start";
    EXEC_STATUS["SUCCESS"] = "success";
    EXEC_STATUS["FAILED"] = "failed";
    EXEC_STATUS["SKIPPED"] = "skipped";
})(EXEC_STATUS = exports.EXEC_STATUS || (exports.EXEC_STATUS = {}));
var ACTIVITY_TYPE;
(function (ACTIVITY_TYPE) {
    ACTIVITY_TYPE["ACTION"] = "action";
    ACTIVITY_TYPE["TASK"] = "task";
    ACTIVITY_TYPE["QUESTION"] = "question";
})(ACTIVITY_TYPE = exports.ACTIVITY_TYPE || (exports.ACTIVITY_TYPE = {}));
