// @testla/screenplay-playwright/api/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Response = exports.Put = exports.Post = exports.Patch = exports.Head = exports.Get = exports.Delete = exports.UseAPI = void 0;
const UseAPI_1 = require("./abilities/UseAPI");
Object.defineProperty(exports, "UseAPI", { enumerable: true, get: function () { return UseAPI_1.UseAPI; } });
const Delete_1 = require("./actions/Delete");
Object.defineProperty(exports, "Delete", { enumerable: true, get: function () { return Delete_1.Delete; } });
const Get_1 = require("./actions/Get");
Object.defineProperty(exports, "Get", { enumerable: true, get: function () { return Get_1.Get; } });
const Head_1 = require("./actions/Head");
Object.defineProperty(exports, "Head", { enumerable: true, get: function () { return Head_1.Head; } });
const Patch_1 = require("./actions/Patch");
Object.defineProperty(exports, "Patch", { enumerable: true, get: function () { return Patch_1.Patch; } });
const Post_1 = require("./actions/Post");
Object.defineProperty(exports, "Post", { enumerable: true, get: function () { return Post_1.Post; } });
const Put_1 = require("./actions/Put");
Object.defineProperty(exports, "Put", { enumerable: true, get: function () { return Put_1.Put; } });
const Response_1 = require("./questions/Response");
Object.defineProperty(exports, "Response", { enumerable: true, get: function () { return Response_1.Response; } });
